mantidproject/mantid
# MultipleFileProperty to allow empty tokens

**Description of work.**
There is a need for SANS to have a possibility to have empty placeholder tokens in MultipleFileProperty that will still pass as valid. See the issue for more details.

This PR implements an option to make this possible.

**To test:**

Build target `SystemTestData`.
Create a fake algorithm with this (edit the `%YOUR_MANTID_BUILD_DIR%` in it):

```
from mantid.simpleapi import *
from mantid.api import *

path = '%YOUR_MANTID_BUILD_DIR%/ExternalData/Testing/Data/SystemTest/ILL/IN16B/'

config.appendDataSearchDir(path)
config['default.facility'] = 'ILL'

class MockAlgorithm(PythonAlgorithm):
    def category(self):
        return 'Test';
    
    def PyInit(self):
        self.declareProperty(MultipleFileProperty('Runs', action = FileAction.OptionalLoad, extensions=['nxs'], allow_empty=True))        
        
    def PyExec(self):
        print(self.getPropertyValue('Runs'))
        
AlgorithmFactory.subscribe(MockAlgorithm)
```

Open the dialog of `MockAlgorithm` and put in `136553,0,136555`.

This should pass the validation and the algorithm should execute and print the full paths of the first and third run, and 000000 for the placeholder in between.

In the contrary, if you change `FileAction.OptionalLoad` to `FileAction.Load`(i.e. mandatory), the validation should fail with the error that the 000000 file is not found.



Fixes #26339 . 

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
